File: src/models/videoCache.js
/**
 * 视频缓存模块 - 异步 I/O 版本
 */
const fs = require('fs').promises;
const fsSync = require('fs');
const path = require('path');

// 缓存数据文件路径
const CACHE_FILE = path.join(__dirname, '../../data/videoCache.json');
const SCAN_LOCK_FILE = path.join(__dirname, '../../data/.scan-lock');

// 内存缓存
let cacheData = {
  videos: [],
  lastScanTime: 0
};
let isScanningFlag = false;

/**
 * 确保数据目录存在
 */
async function ensureDataDir() {
  const dataDir = path.dirname(CACHE_FILE);
  await fs.mkdir(dataDir, { recursive: true }).catch(err => {
    if (err.code !== 'EEXIST') throw err;
  });
}

/**
 * 读取缓存数据
 */
async function readCache() {
  try {
    const data = await fs.readFile(CACHE_FILE, 'utf8');
    cacheData = JSON.parse(data);
    return cacheData;
  } catch (err) {
    if (err.code === 'ENOENT') {
      cacheData = { videos: [], lastScanTime: 0 };
      return cacheData;
    }
    console.error('读取视频缓存失败:', err);
    return cacheData;
  }
}

/**
 * 保存缓存数据
 */
async function saveCache(data) {
  await ensureDataDir();
  try {
    await fs.writeFile(CACHE_FILE, JSON.stringify(data, null, 2), 'utf8');
  } catch (err) {
    console.error('保存视频缓存失败:', err);
  }
}

/**
 * 获取视频列表（同步，从内存返回）
 * @returns {Array} 视频列表
 */
function getVideos() {
  return cacheData.videos || [];
}

/**
 * 设置视频列表（异步）
 * @param {Array} videos 视频列表
 */
async function setVideos(videos) {
  cacheData.videos = videos;
  cacheData.lastScanTime = Date.now();
  await saveCache(cacheData);
}

/**
 * 检查是否正在扫描
 * @returns {boolean}
 */
function isScanning() {
  if (!isScanningFlag) return false;
  
  // 检查锁是否超时（10分钟）
  if (!fsSync.existsSync(SCAN_LOCK_FILE)) {
    isScanningFlag = false;
    return false;
  }
  
  try {
    const lockData = JSON.parse(fsSync.readFileSync(SCAN_LOCK_FILE, 'utf8'));
    if (Date.now() - lockData.lockedAt > 10 * 60 * 1000) {
      fsSync.unlinkSync(SCAN_LOCK_FILE);
      isScanningFlag = false;
      return false;
    }
    return true;
  } catch (err) {
    return false;
  }
}

/**
 * 设置扫描状态
 * @param {boolean} scanning 是否正在扫描
 */
async function setScanning(scanning) {
  await ensureDataDir();
  isScanningFlag = scanning;
  
  if (scanning) {
    await fs.writeFile(SCAN_LOCK_FILE, JSON.stringify({
      lockedAt: Date.now()
    }), 'utf8');
  } else {
    try {
      await fs.unlink(SCAN_LOCK_FILE);
    } catch (err) {
      // 忽略文件不存在的错误
    }
  }
}

/**
 * 从文件加载缓存（启动时调用）
 * @returns {Promise<boolean>} 是否加载成功
 */
async function loadCache() {
  const cache = await readCache();
  
  if (cache.videos && cache.videos.length > 0) {
    console.log(`已加载视频缓存: ${cache.videos.length} 个视频`);
    return true;
  }
  
  return false;
}

/**
 * 获取缓存状态
 * @returns {object}
 */
function getStatus() {
  return {
    isScanning: isScanningFlag,
    videoCount: (cacheData.videos || []).length,
    lastScanTime: cacheData.lastScanTime || 0
  };
}

/**
 * 更新最后扫描时间
 */
async function updateLastScanTime() {
  cacheData.lastScanTime = Date.now();
  await saveCache(cacheData);
}

module.exports = {
  getVideos,
  setVideos,
  setScanning,
  isScanning,
  getStatus,
  updateLastScanTime,
  loadCache,
  saveCache
};

File: go-impl/internal/model/videoCache.go
// Package model 数据模型
package model

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"sync/atomic"
	"time"
)

// Video 视频信息
type Video struct {
	Name         string `json:"name"`
	RelativePath string `json:"relativePath"`
	FullPath     string `json:"fullPath"`
	DirName      string `json:"dirName"`
	Size         string `json:"size"`
	SizeBytes    int64  `json:"sizeBytes"`
	Modified     string `json:"modified"`
	ModifiedTime int64  `json:"modifiedTime"`
	ID           string `json:"id"` // 预计算的ID
}

// CacheData 缓存数据
type CacheData struct {
	Videos       []Video `json:"videos"`
	LastScanTime int64   `json:"lastScanTime"`
}

// VideoCache 视频缓存
type VideoCache struct {
	filePath   string
	lockPath   string
	data       atomic.Pointer[CacheData]
	isScanning atomic.Bool

	// 预序列化的JSON缓存
	jsonCache atomic.Pointer[[]byte]
}

var (
	videoCache *VideoCache
	cacheOnce  sync.Once
)

// GetVideoCache 获取视频缓存单例
func GetVideoCache() *VideoCache {
	cacheOnce.Do(func() {
		execPath, _ := os.Executable()
		baseDir := filepath.Dir(execPath)
		videoCache = &VideoCache{
			filePath: filepath.Join(baseDir, "data", "videoCache.json"),
			lockPath: filepath.Join(baseDir, "data", ".scan-lock"),
		}
		videoCache.data.Store(&CacheData{Videos: []Video{}})
	})
	return videoCache
}

// LoadCache 加载缓存
func (vc *VideoCache) LoadCache() bool {
	fileData, err := os.ReadFile(vc.filePath)
	if err != nil {
		return false
	}

	var data CacheData
	if err := json.Unmarshal(fileData, &data); err != nil {
		return false
	}

	if len(data.Videos) > 0 {
		// 为每个视频计算ID
		for i := range data.Videos {
			if data.Videos[i].ID == "" {
				data.Videos[i].ID = GenerateVideoID(data.Videos[i].DirName, data.Videos[i].RelativePath)
			}
		}
		vc.data.Store(&data)
		return true
	}
	return false
}

// GenerateVideoID 生成视频ID
func GenerateVideoID(dirName, relativePath string) string {
	data := dirName + ":" + relativePath
	hash := sha256.Sum256([]byte(data))
	return hex.EncodeToString(hash[:])[:16]
}

// GetVideos 获取视频列表
func (vc *VideoCache) GetVideos() []Video {
	data := vc.data.Load()
	if data == nil {
		return []Video{}
	}
	return data.Videos
}

// SetVideos 设置视频列表
func (vc *VideoCache) SetVideos(videos []Video) {
	data := &CacheData{
		Videos:       videos,
		LastScanTime: time.Now().UnixMilli(),
	}
	vc.data.Store(data)

	// 清除JSON缓存
	vc.jsonCache.Store(nil)

	// 异步保存
	go func() {
		os.MkdirAll(filepath.Dir(vc.filePath), 0755)
		jsonData, _ := json.MarshalIndent(data, "", "  ")
		os.WriteFile(vc.filePath, jsonData, 0644)
	}()
}

// IsScanning 是否正在扫描
func (vc *VideoCache) IsScanning() bool {
	if !vc.isScanning.Load() {
		return false
	}

	data, err := os.ReadFile(vc.lockPath)
	if err != nil {
		vc.isScanning.Store(false)
		return false
	}

	var lockData struct {
		LockedAt int64 `json:"lockedAt"`
	}
	if json.Unmarshal(data, &lockData) != nil {
		vc.isScanning.Store(false)
		return false
	}

	if time.Now().UnixMilli()-lockData.LockedAt > 10*60*1000 {
		os.Remove(vc.lockPath)
		vc.isScanning.Store(false)
		return false
	}

	return true
}

// SetScanning 设置扫描状态
func (vc *VideoCache) SetScanning(scanning bool) {
	vc.isScanning.Store(scanning)
	os.MkdirAll(filepath.Dir(vc.lockPath), 0755)

	if scanning {
		lockData, _ := json.Marshal(map[string]int64{"lockedAt": time.Now().UnixMilli()})
		os.WriteFile(vc.lockPath, lockData, 0644)
	} else {
		os.Remove(vc.lockPath)
	}
}

// GetStatus 获取状态
func (vc *VideoCache) GetStatus() map[string]interface{} {
	data := vc.data.Load()
	videoCount := 0
	lastScanTime := int64(0)
	if data != nil {
		videoCount = len(data.Videos)
		lastScanTime = data.LastScanTime
	}

	return map[string]interface{}{
		"isScanning":   vc.isScanning.Load(),
		"videoCount":   videoCount,
		"lastScanTime": lastScanTime,
	}
}
